src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -0,0 +1,385 @@
+                Train(const DataSet& dataset, const Config& cfg) override {
+                    auto ivf_raft_cfg = static_cast<const typename KnowhereConfigType<T>::Type&>(cfg);
+                    auto result = Status::success;
+                    if (gpu_index_) {
+                        LOG_KNOWHERE_WARNING_ << "index is already trained";
+                        result = Status::index_already_trained;
+                    } else if (ivf_raft_cfg.gpu_ids.size() == 1) {
+                        try {
+                            auto metric = Str2RaftMetricType(ivf_raft_cfg.metric_type);
+                            if (!metric.has_value()) {
+                                LOG_KNOWHERE_WARNING_ << "please check metric value: " << ivf_raft_cfg.metric_type;
+                                return metric.error();
+                            }
+                            if (
+                                    metric.value() != raft::distance::DistanceType::L2Expanded
+                                    && metric.value() != raft::distance::DistanceType::L2Unexpanded
+                                    && metric.value() != raft::distance::DistanceType::InnerProduct
+                               ) {
+                                LOG_KNOWHERE_WARNING_ << 
+                                    "selected metric not supported in RAFT IVF indexes: " << 
+                                    ivf_raft_cfg.metric_type;
+                                return Status::invalid_metric_type;
+                            }
+
+                            auto scoped_device = detail::device_setter{
+                                *ivf_raft_cfg.gpu_ids.begin()
+                            };
+                            res_ = std::make_unique<raft::device_resources>();
+                            auto rows = dataset.GetRows();
+                            auto dim = dataset.GetDim();
+                            auto* data = reinterpret_cast<float const*>(dataset.GetTensor());
+
+                            auto stream = res_->get_stream();
+                            auto data_gpu = rmm::device_uvector<float>(rows * dim, stream);
+                            RAFT_CUDA_TRY(
+                                    cudaMemcpyAsync(
+                                        data_gpu.data(),
+                                        data,
+                                        data_gpu.size() * sizeof(float),
+                                        cudaMemcpyDefault,
+                                        stream.value()
+                                        )
+                                    );
+                            if constexpr (std::is_same_v<detail::raft_ivf_flat_index, T>) {
+                                auto build_params = raft::neighbors::ivf_flat::index_params{};
+                                build_params.metric = metric.value();
+                                build_params.n_lists = ivf_raft_cfg.nlist;
```

Comment:
I'm assuming there's going to be a parameter sweep in order to tune the settings for computing perf at different recall levels? We should expose all of the params so we can sweep through as much of them as possible- there aren't as many for ivf flat as ivf pq, but it will probably be helpful to expose the number of coarse quantizer iterations and amount of the dataset to use for training the coarse quantizer here.

Reply:
As long as this first implementation is not assumed to be "benchmarking" complete (e.g. any benchmarking will not be assumed valid) then I think that's fine. Otherwise, at the very least, the additional ivf_pq parameters are going to be needed to achiece competitive performance.

Reply:
Yes, I think you're right. Given that we're going to want to move fairly quickly toward benchmarking this, maybe we'd better expose that sooner rather than later. I'll have a chat with other folks and see how best to pass through the necessary parameters.